src/helpers.rs

```diff
@@ -512,6 +516,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let value_place = op_place.offset(offset, MemPlaceMeta::None, layout, this)?;
         this.write_scalar(value, value_place.into())
     }
+
+    /// Parse a `timespec` struct and return it as a `std::time::Duration`. The outer `Result` is
+    /// for interpreter errors encountered while reading memory, and the inner `Result` indicates
+    /// whether the value in the `timespec` struct is invalid. Some libc functions will return
+    /// `EINVAL` if the struct's value is invalid.
+    fn read_timespec(
+        &mut self,
+        timespec_ptr_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, Result<Duration, TimespecError>> {
+        let this = self.eval_context_mut();
+        let tp = this.deref_operand(timespec_ptr_op)?;
+        let seconds_place = this.mplace_field(tp, 0)?;
+        let seconds_scalar = this.read_scalar(seconds_place.into())?;
+        let seconds = seconds_scalar.to_machine_isize(this)?;
+        let nanoseconds_place = this.mplace_field(tp, 1)?;
+        let nanoseconds_scalar = this.read_scalar(nanoseconds_place.into())?;
+        let nanoseconds = nanoseconds_scalar.to_machine_isize(this)?;
+
+        let seconds: u64 = if let Ok(s) = seconds.try_into() {
+            s
+        } else {
+            return Ok(Err(TimespecError));
```

Comment:
Is this something that can ever fail? These are <=64bit fields on all platforms, right, so we could also just `unwrap` here?

Reply:
Oh, it can fail if it is negative, right? Makes sense, please add a comment.